crates/solver/src/settlement_submission/submitter.rs

```diff
@@ -493,10 +487,14 @@ impl<'a> Submitter<'a> {
             // execute transaction
 
             let label: &'static str = self.submit_api.name().into();
-            match self.submit_api.submit_transaction(method.tx).await {
+            let (hash, raw) = match method.tx.build().now_or_never().unwrap().unwrap() {
+                Transaction::Request(_) => unreachable!(),
+                Transaction::Raw { bytes, hash } => (hash, RawTransaction(bytes.0)),
+            };
+            transactions.push((hash, gas_price));
+            match self.submit_api.submit_transaction(raw).await {
                 Ok(handle) => {
                     tracing::debug!(?handle, "submitted transaction",);
-                    transactions.push((handle, gas_price));
```

Comment:
Moving this push up fixes the race condition.

Reply:
I think if the tx submission actually fails we will not try to re-submit it with this change. Would it also fix the race condition if this tx gets removed from `transactions` in case we detect an error?

Reply:
If we remove in the other-error case, then the following can happen:
- We add hash to transactions.
- Submission errors.
- We remove hash from transactions.
- Submission didn't actually error. We just got a random http error.
- Transactions gets mined but we treat the submission as failed because the hash isn't in transactions.

Reply:
The moment we initiate a tx submit, we need to keep the hash in transactions. That's the only reliable way to not miss a transaction. I don't even trust nodes to not accidentally add the new tx to their mempool even if they reply with "already known".

Reply:
> Transactions gets mined but we treat the submission as failed because the hash isn't in transactions.

Hmm, that's true. Maybe we should just don't care that we already tried to submit the tx. Realistically we will only submit a tx every 2s (I think) so we wouldn't overwhelm a node by dropping this optimization but we might make the mining logic more reliable.

Reply:
Yeah I think what we should do is always try to submit and if we get "tx already known error" because gas price hasn't increased enough ignore it. I would like to do that in another PR though.